### PR TITLE
Recommendation 3.1 - recommended order of imports

### LINKED ISSUE #55: Recommendation 3.1 - recommended ordering of imports
The following snippet can be a starting point for implementation inside `FileStructureRule.kt`:

```
private val recommendedComparator = Comparator<ASTNode> { import1, import2 ->
        require(import1.elementType == IMPORT_DIRECTIVE && import2.elementType == IMPORT_DIRECTIVE) { "This comparator is for sorting imports" }
        val pathSegments1 = (import1.psi as KtImportDirective).importPath!!.fqName.pathSegments()
        val pathSegments2 = (import2.psi as KtImportDirective).importPath!!.fqName.pathSegments()

        return@Comparator 0
    }
```

### PR DESCRIPTION
**Pull request by petertrr**
## Which rule and warnings did you add?

Recommendation 3.1 - imports grouping

This pull request closes #55

## Actions checklist
* [x] Implemented Rule, added Warnings
* [x] Added tests on checks
* [x] Added tests on fixers
* [x] Updated diktat-analysis.yml
* [ ] Updated available-rules.md

### CODE REVIEW

File: diktat-analysis.yml [hunk 1]
```diff
@@ -146,7 +147,8 @@
   configuration: {}
 - name: FILE_WILDCARD_IMPORTS
   enabled: true
-  configuration: {}
+  configuration:
+    allowedWildcards: "" # Allowed wildcards for imports (e.g. "import org.cqfn.diktat.*, import org.jetbrains.kotlin.*")
```
Comment (aktsay6) [inline diktat-analysis.yml:151]: Do we need to have useRecommendedImportsOrder like in others yml config files?

Comment (petertrr) [reply diktat-analysis.yml:151]: Yes, you are right, I'll add it

Comment (aktsay6) [reply diktat-analysis.yml:151]: And what if someone will want to sort wildcart imports and don't want to sort non-wildcard?

Comment (petertrr) [reply diktat-analysis.yml]: I don't think there is much profit in separating wildcard and non-wildcard imports in this case. For example, IDEA doesn't separate them.

File: info/diktat-kotlin-coding-style-guide-en.md [hunk 2]
```diff
@@ -656,56 +656,34 @@ b) Each of the above code blocks should be separated by a blank line.
...
 From top to bottom, the order is the following:
-1. static imports
-2. Android
-3. Imports of packages used internally in your organization
-4. Other commercial organizations
-5. Open source imports
-6. net/org open source dependencies
-7. javacard
-8. java core packages
+1. Android
```
Comment (akuleshov7) [inline info/diktat-kotlin-coding-style-guide-en.md:662]: much better now

File: diktat-rules/src/test/resources/test/paragraph3/file_structure/ReorderingImportsRecommendedTest.kt [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package test.paragraph3.file_structure
+
```
Comment (akuleshov7) [inline diktat-rules/src/test/resources/test/paragraph3/file_structure/ReorderingImportsRecommendedTest.kt:2]: need to add more third party libs

File: diktat-rules/src/main/resources/diktat-analysis.yml [hunk 4]
```diff
@@ -138,7 +138,8 @@
   configuration: {}
 - name: FILE_UNORDERED_IMPORTS
   enabled: true
-  configuration: {}
+  configuration:
```
Comment (akuleshov7) [inline diktat-rules/src/main/resources/diktat-analysis.yml:141]: please add description

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt [hunk 5]
```diff
@@ -107,56 +125,127 @@ class FileStructureRule(private val configRules: List<RulesConfig>) : Rule("file
...
+        if (sortedImportsGroups.flatten() != imports) {
             FILE_UNORDERED_IMPORTS.warnAndFix(configRules, emitWarn, isFixMode, fileName, node.startOffset, node) {
-                rearrangeImports(node, imports, sortedImports)
+                rearrangeImports(node, imports, sortedImportsGroups)
             }
         }
     }
 
-    private fun rearrangeImports(node: ASTNode, imports: List<ASTNode>, sortedImports: List<ASTNode>) {
+    private fun rearrangeImports(node: ASTNode, imports: List<ASTNode>, sortedImportsGroups: List<List<ASTNode>>) {
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:159]: can be a set of sets

File: diktat-analysis.yml (hunk 1 above)
Comment (aktsay6) [reply diktat-analysis.yml]: Are you using `useRecommendedImportsOrder` from `FILE_WILDCARD_IMPORTS` anywhere?

Comment (petertrr) [reply diktat-analysis.yml]: I already moved it to `FILE_UNORDERED_IMPORTS`, it was initially by mistake in `FILE_WILDCARD_IMPORTS `

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt [hunk 6]
```diff
@@ -52,10 +58,14 @@ class FileStructureRule(private val configRules: List<RulesConfig>) : Rule("file
 
         if (node.elementType == ElementType.FILE) {
             fileName = node.getFileName()
-            val configuration = WildCardImportsConfig(
-                    this.configRules.getRuleConfig(FILE_WILDCARD_IMPORTS)?.configuration ?: mapOf()
+            val wildcardImportsConfig = WildCardImportsConfig(
+                this.configRules.getRuleConfig(FILE_WILDCARD_IMPORTS)?.configuration ?: mapOf()
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:62]: emptyMap

File: diktat-analysis.yml (hunk 1 above)
Comment (aktsay6) [reply diktat-analysis.yml]: Ok

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt [hunk 7]
```diff
@@ -107,56 +125,127 @@ class FileStructureRule(private val configRules: List<RulesConfig>) : Rule("file
...
+    private data class ImportGroups(
+        val android: List<KtImportDirective>,
+        val ownDomain: List<KtImportDirective>,
+        val others: List<KtImportDirective>,
+        val javaCore: List<KtImportDirective>,
+        val kotlinCore: List<KtImportDirective>
+    ) {
+        fun toList() = listOf(android, ownDomain, others, javaCore, kotlinCore)
+
+        companion object {
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:206]: oh, I do not like these calculations in companion object. Why it should be here?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt [hunk 8]
```diff
@@ -107,56 +125,127 @@ class FileStructureRule(private val configRules: List<RulesConfig>) : Rule("file
...
+        fileAnnotations: ASTNode?,
+        packageDirectiveNode: ASTNode
     ): Pair<ASTNode?, ASTNode> = when (elementType) {
         BLOCK_COMMENT -> null to listOfNotNull(headerKdoc, fileAnnotations, packageDirectiveNode).first()
         KDOC -> copyrightComment to (fileAnnotations ?: packageDirectiveNode)
         FILE_ANNOTATION_LIST -> (headerKdoc ?: copyrightComment) to packageDirectiveNode
         else -> error("Only BLOCK_COMMENT, KDOC and FILE_ANNOTATION_LIST are valid inputs.")
     }
 
+    private data class ImportGroups(
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:197]: you are not using anything from a data class here 
simply generating extra code

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt [hunk 9]
```diff
@@ -107,56 +125,127 @@ class FileStructureRule(private val configRules: List<RulesConfig>) : Rule("file
...
+            }
+
+            @Suppress("UnsafeCallOnNullableType")
+            private inline fun List<KtImportDirective>.partitionByImportParts(predicate: (List<Name>) -> Boolean) =
+                partition {
+                    it.importPath!!.fqName.pathSegments().let(predicate)
+                }
+
+            private fun List<Name>.isStandardAndroid() =
+                first().identifier.let { it == "android" || it == "androidx" } ||
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:232]: oh, need to move these to constants

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt (hunk 7 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:206]: I wanted to place a factory method inside the class itself. I sure can extract it to the Rule class level, but why is placing it in `companion object` bad?

**Review by akuleshov7 — APPROVED** [review]
lgtm

**Review by aktsay6 — APPROVED** [review]
lgtm

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt (hunk 5 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:159]: Why? They are already sorted

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt (hunk 8 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:197]: Yes, I must've thought that eventually there will be more grouping-specific logic. I'll remove data class.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt (hunk 9 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:232]: I'd better extract lists of strings to `ASTNodeConstants`.

**Pull request merged**